Create edit_reports permission for configurable reports

solleks opened the pull request:
This is a copy of an [earlier draft PR](https://github.com/dimagi/commcare-hq/pull/27993) that I messed up due to my ineptness with git.

[Feature request document.](https://docs.google.com/document/d/1bmo7swOiXyUTW6IcWn9HFCNcdUS8m6KzMcSK2biN7zg/edit?ts=5ef3962a)

[Trello card.](https://trello.com/c/YjJacM4G/83-detach-reports-from-data-permissions)

##### SUMMARY
This change creates a new permission ‘edit_reports’ that controls two actions - creating custom reports and editing custom reports. The pieces of the change are:

- Add the ‘edit_reports’ property to the corehq.apps.users.models.Permissions class.
- Add a migration for the new property that copies the ‘edit_data’ property.
- Change the following permissions checks from ‘edit_data’ to ‘edit_reports’
    - Whether the “Report Builder” section appears in the ProjectReportsTab
    - The method_decorator on ReportBuilderView.dispatch
- Propagate ‘edit_reports’ to page_context of ConfigurableReportView to avoid showing the ‘Edit Report’ button.
- Add ‘Reports’ section to edit role modal dialog, which contains Create and Edit Reports checkbox for ‘edit_reports’ permission and existing Access All Reports and Access Specific Reports controls.
<img width="452" alt="EditRole" src="https://user-images.githubusercontent.com/1011180/86085930-40c72180-ba55-11ea-9a80-1ebdbc084cc4.png">
- Change Reports column of project ‘settings/users/roles/’ page to show a check mark when role has ‘edit_reports’ permission and a view only icon when role has only ‘view_reports’ permission.
<img width="611" alt="RolePermissions" src="https://user-images.githubusercontent.com/1011180/86085952-51779780-ba55-11ea-97c2-deacb19fb029.png">

##### FEATURE FLAG
Not specific to a feature flag.

##### RISK ASSESSMENT / QA PLAN
This change should get a QA plan as it will affect users permissions to take actions.
I need to learn how to set up a QA plan.

##### PRODUCT DESCRIPTION
This change allows project administrators to create roles that can edit data but not create/edit reports and vice versa. They will make the choice in the edit role dialog and see the effect in the role and permissions page.

solleks committed:
Put Report related permissions in their own section of the edit role modal

solleks committed:
Add Create and Edit Reports checkbox

solleks committed:
Create edit_reports permission for configurable reports

solleks committed:
Check for new edit_reports permission

solleks committed:
Show view only and edit icons for Reports

solleks committed:
Add a test for new edit_reports permission

solleks committed:
Restore deleted blank line

solleks commented:
proteusvacuum reviewed my earlier draft PR of this change. I will copy his comments into this version.

solleks commented:
From proteusvacuum:

Did you mean to commit the submodule changes? Sometimes these diverge from the branch you are working on, and it doesn't seem like the changes in those are related to this PR.

To make a QA request you can fill out this form: https://docs.google.com/forms/d/e/1FAIpQLSehQl5VXuHED_HLXaMcUoWleBGqpZu_chPXfWaxdcb0QbDh4Q/formResponse and the QA team will reach out to you if they need clarifications.

We normally deploy QA branches to staging - instructions for how to do that are found in scripts/staging.yaml, but feel free to reach out in #staging on slack if you have questions about that, I'm not sure if you have deploy permissions on the staging machines.

Feel free to reach out to me on slack if you wanted to chat through anything else, although this seems pretty complete already.

solleks commented:
The submodule changes were not intentional, so I removed them.

I don't have permission to open the QA request form. Cal suggested that I contact Ben Talbot, so I will do that separately.

I'm sure that I don't have permissions to touch anything in staging.

solleks commented on corehq/apps/users/migrations/0019_editreportspermissions.py:
```diff
@@ -0,0 +1,38 @@
+        if role.permissions.edit_data:
+            role.permissions.edit_reports = True
+            changed = True
+        elif role.permissions.edit_reports:
```
proteusvacuum commented "I'm not sure I follow why this is being set to False here."

I borrowed this logic from the web apps permission change, which was added after the initial change:
https://github.com/dimagi/commcare-hq/blame/master/corehq/apps/users/migrations/0016_webappspermissions.py

It makes sure that edit_reports won't be true if edit_data is false.

solleks commented on corehq/apps/users/models.py:
```diff
@@ -2708,6 +2710,10 @@ def get_id(self):
     def is_active(self):
         return True
 
+    @property
```
I added this so that a unit test would work, but I want to call attention to this change to make sure it's alright.

esoergel replied:
This should be fine, though I don't think the tests should be using `AnonymousCouchUser`s (more about that above)

solleks committed:
Test edit_reports permission with WebUser instead of AnonymousCouchUser

solleks committed:
Fix indentation

solleks commented on corehq/apps/userreports/tests/test_view.py:
```diff
@@ -207,3 +209,30 @@ def test_redirect_report_builder(self):
+
+        class FakeCouchUser(AnonymousCouchUser):
+            def can_edit_reports(self):
+                return True
```
Thanks for the pointers. I've changed the test to use a WebUser instead of the AnonymousCouchUser.

esoergel commented on corehq/apps/userreports/tests/test_view.py:
```diff
@@ -207,3 +211,51 @@ def test_redirect_report_builder(self):
+        Test whether ConfigurableReportView.page_context allows report editing
+        """
+        domain = Domain(name='test_domain', is_active=True)
+        domain.save()
```
Can you also make sure the test cleans up these objects?  `self.addCleanup(domain.delete)` is probably the easiest way to do it.  That should also cascade delete the roles, but the web users will need to be deleted individually.

solleks replied:
Sure, I've added those calls to addCleanup.

esoergel commented on corehq/apps/userreports/tests/test_view.py:
```diff
@@ -207,3 +211,51 @@ def test_redirect_report_builder(self):
+
+            request = HttpRequest()
+            request.can_access_all_locations = True
+            request.user = web_user
```
```suggestion
            request.user = web_user.get_django_user()
```

solleks replied:
Done.

solleks committed:
Delete DB objects created during test

solleks commented:
Could you add the label "product/all-users-all-environments" to this request? I don't seem to be able to add a label.

calellowitz commented on corehq/apps/users/templates/users/roles_and_permissions.html:
```diff
@@ -164,9 +164,9 @@ <h6 data-bind="visible: !name()">({% trans "No Name" %})</h6>
-            <div data-bind="visible: reportPermissions.all(),
-                                            permissionIcon: {
-                                                manage: true
+            <div data-bind="permissionIcon: {
```
It may not be necessary to add this to the default page. Specifically, we appear to generally exclude permissions that require a specific subscription level, like this one, since they are confusing for projects that do not have it. Other new permissions like the web apps one are likewise excluded from this page.

solleks replied:
Let me make sure I understand. I don't think you're saying we should remove the Reports column from this table because it was there before this change. Rather, you're saying that I should just undo this change so the column shows either a check mark, a list of specific reports or empty, but doesn't distinguish the ability to edit from the ability to view. Does that sound right?

calellowitz replied:
Ah, that is an excellent catch, and I definitely missed that it was there before. After this change is it possible for a user to be able to be able to create reports, but not view all reports? What is shown if they can only see some of the reports? It looks like the code that lists reports is still there, do they see the eye and the list? That might be crowded. Depending on the answer to the first question, I think the original behavior is fine, or if its not possible to edit but be restricted, having icons for edit and view all, and the list for restricted seems fine as well.

solleks replied:
It is possible for a user to be able to create reports but not view all reports. In that case, they would see both a check mark and a list of reports. Removing the ability to create reports makes it show just the list, with no eye icon (not really sure why). Giving access to all reports without the ability to create them does show the eye icon with "view only".

I propose that we restore this to the way it was - a check mark when users can access all reports and a list when they can only access specific ones. Since the builtin reports are available with any subscription, this is consistent with your first comment. The ability to access reports is available to all so it makes sense in this table, while the ability to create reports requires pro or higher subscriptions so we won't try to show it in this table.

calellowitz commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -511,6 +511,64 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
+            </legend>
+
+            <div class="form-group">
+              <label class="col-sm-4 control-label">
```
I think this should be hidden if the user's subscription does not have access to the report builder feature

solleks replied:
How can we implement that check?

calellowitz replied:
It looks like this util function does so https://github.com/dimagi/commcare-hq/blob/a37dc36631dbff5f5c63a9d9b49b83a16ded1221/corehq/apps/userreports/util.py#L49

solleks replied:
Thanks for the pointer. I've added has_report_builder_access to the context and conditionalized this item in the dialog.

calellowitz reviewed (approved):
Generally this looks great (and the screenshots look good as well). Few small comments but thanks for doing this

solleks committed:
Protect delete_report view. Don't show edit_reports in user role table.

solleks commented on corehq/apps/userreports/views.py:
```diff
@@ -309,7 +309,7 @@ class CreateConfigReportView(BaseEditConfigReportView):
 class ReportBuilderView(BaseDomainView):
 
-    @method_decorator(require_permission(Permissions.edit_data))
+    @method_decorator(require_permission(Permissions.edit_reports))
```
OK, I've added the checks that make sense to me:
    o login and domain required
    o UCR toggle set to on
    o has edit_reports permission

solleks commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -511,6 +511,64 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
+
+            <div class="form-group">
+              <label class="col-sm-4 control-label">
+                {% trans "Create and Edit Reports" %}
```
I've added "in report builder" to the label in the form-check (i.e. on the right) but left it out of the label on the form-group because it wrapped and looked out of place with the other labels in its column.

solleks committed:
Hide Create/Edit Reports setting unless role has full organization access

solleks commented:
I added one small change to hide the Create and Edit Reports setting when Full Organization Access is deselected. Location-restricted users aren't able to create custom reports now with 'edit_data' and won't be able to with 'edit_reports' either, and hiding the setting makes this more apparent.

solleks commented:
I made some changes after Cal's approval and I'd be happy to get feedback on those if you have any.

One QA pass has finished successfully, but Ben said he would be doing another pass. Once that is done and assuming no issues arise with my post-approval changes, I will ask one of you reviewers to perform the merge.

The [documentation](https://confluence.dimagi.com/display/commcarepublic/Roles+and+Permissions) is updated in draft form if anyone wants to take look. That includes adding a description of the Web Apps permission.

solleks commented:
The QA review has completed with no further issues (https://dimagi-dev.atlassian.net/browse/QA-1515).

Cal or Ethan, if my most recent changes look good to you, could you perform the merge? I do not have permission to do it.

calellowitz commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -516,7 +516,7 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
 
             {% if has_report_builder_access %}
-              <div class="form-group">
+              <div class="form-group" data-bind="visible: permissions.access_all_locations">
```
If a user removes full location access, it looks like this permission will still remain. I think if we need to restrict this only to roles with full access, then we will need to make sure to reset this permission when the other is removed

solleks replied:
I agree that it should probably behave the way you suggest. But it also looks like the existing permissions that depend on access_all_locations (e.g. Web Users and Groups) are not reset either.

I don't really know Knockout.js, so I'd be reluctant to attempt this in the dialog box itself. It looks like it could be done in the 'post_user_role' function though.

solleks replied:
I added code to disable permissions for location-restricted users. There was a test failure due to new arguments in a method, so I've synced my branch and added a fix.

calellowitz reviewed (changes requested):
Thanks for making those changes. One thought about the last commit, but otherwise looks good

solleks committed:
Disable incompatible permissions for location-restricted roles

solleks committed:
Merge remote-tracking branch 'upstream/master' into cdg/edit_reports_permission

Sync with master to avoid test failures when deleting users.

solleks committed:
Add deleted_by argument to test

solleks committed:
Add permissions checks to undelete_report

solleks commented:
I added permissions checks to the undelete_report view to match delete_view. I also removed the decorator to require UCR because the function _assert_report_delete_privileges is already used.

calellowitz reviewed (approved):
Looks good! Thanks for the followup

solleks commented:
Thanks. And I will need you or someone else with write access to perform the merge for me.
